### PR TITLE
[fix] oss dict load fix

### LINKED ISSUE #380: [OSS] When loading a state, the insertion order for the params may not match the params_groups
## 🐛 Bug
(FB Only f250383189) Guess is that the issue comes from the parameters not being inserted in the right order in the sharded optimizer state. [See](https://github.com/facebookresearch/fairscale/blob/master/fairscale/optim/oss.py#L403)

Would need fixing + repro in a unit test

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Fix for an issue that Pyspeech was seeing, it looks like I broke the logic in the last updates of  #310
The unit test did not catch that because the model was symmetrical enough, tryign to find a better test 
The bug detection and fix suggestion is from Weiyi Zheng @zhengwy888

Fixes #380

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: fairscale/optim/oss.py [hunk 1]
```diff
@@ -391,16 +391,16 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
...
-        for i_param, (key, value) in enumerate(state_dict["state"].items()):
-            param = self.index_to_param[i_param]
+        for key, value in state_dict["state"].items():
+            param = self.index_to_param[pytorch15_index_redirect[key]]
 
             # Populate the sharded optimizer state on the fly
             if self.param_to_rank[param] != self.rank:
                 state_dict["state"][key] = None
 
-            if key in self.index_to_param:
```
Comment (blefaudeux) [inline fairscale/optim/oss.py:402]: this test was definitely wrong, I should have seen that..

File: fairscale/optim/oss.py [hunk 2]
```diff
@@ -391,16 +391,16 @@ def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
 
         # NOTE: PyTorch 1.5 does not index linearly but with the id(params) at saving time
         # we work around that here by using the fact that the params are ordered as in the param_groups
+        pytorch15_index_redirect = {k: i for i, k in enumerate(state_dict["state"].keys())}
```
Comment (blefaudeux) [inline fairscale/optim/oss.py:394]: this is useless post pytorch 1.5

**Comment by blefaudeux** [comment]
context: Planning for a follow up PR to find a unit test which exposes the previous bug (guess is that the model used was too symmetrical), but trying to fix master ASAP in the meantime

**Review by min-xu-ai — APPROVED** [review]
nice!

**Pull request merged**